src/typeInference/annotateBinaryOperator.ts

```diff
@@ -1,165 +1,108 @@
-import isSubtypeOf from '../expressions/dataTypes/isSubtypeOf';
-import { SequenceType, ValueType } from '../expressions/dataTypes/Value';
+import { SequenceType, sequenceTypeToString, ValueType } from '../expressions/dataTypes/Value';
+import { addDuration as addDurationToDateTime } from '../expressions/dataTypes/valueTypes/DateTime';
+import { add as dayTimeDurationAdd } from '../expressions/dataTypes/valueTypes/DayTimeDuration';
+import { add as yearMonthDurationAdd } from '../expressions/dataTypes/valueTypes/YearMonthDuration';
 import { IAST } from '../parsing/astHelper';
+import { BinaryEvaluationFunction } from './binaryEvaluationFunction';
 import { insertAttribute } from './insertAttribute';
 
-type BinOpLookupTable = {
-	[key: number]: ValueType;
+function hash(left: ValueType, right: ValueType, op: string): number {
+	return (
+		((left as number) << 20) +
+		((right as number) << 12) +
+		(op.charCodeAt(0) << 8) +
+		op.charCodeAt(1)
```

Comment:
Shifting is a cool idea, if typescript manages to keep up the performance and does not do some weird things

Reply:
I think it will work for now but this might have to be changed in the future.